### PR TITLE
Refactor foreman_setting module to apypie

### PR DESCRIPTION
**Pull request by mdellweg**
Also, remove deprecated foreman_setting_facts module.

### CODE REVIEW

**Comment by mdellweg** [comment]
@sean797 I needed to add some strange workaround for resetting to a default for lists.
Can you please have a look. I think you have more insight on the type-checking matter.

File: plugins/modules/foreman_setting.py [hunk 1]
```diff
@@ -92,32 +78,31 @@
...
-        entity_dict['value'] = entity.default or ""
-
-    entity_dict = sanitize_entity_dict(entity_dict, name_map)
+        default = entity['default'] or ""
+        if isinstance(default, list):
+            default = ','.join(default)
+        entity_dict['value'] = default
 
-    changed, entity = naildown_entity(Setting, entity_dict, entity, 'present', module, check_type=True)
+    # add state='present' as soon as it is available again
```
Comment (mdellweg) [inline plugins/modules/foreman_setting.py]: This would be when #337 is ready.

File: plugins/modules/foreman_setting_facts.py [hunk 2]
```diff
@@ -1,162 +0,0 @@
-#!/usr/bin/python
```
Comment (evgeni) [inline plugins/modules/foreman_setting_facts.py]: is this removal intentional?

Comment (evgeni) [reply plugins/modules/foreman_setting_facts.py]: disregard this comment, it is according to the initial PR comment

**Comment by mdellweg** [comment]
@evgeni @sean797 
This changes the original interface (which we never stated to be stable...) in that lists should be passed as lists now, and order matters.

File: tests/test_playbooks/setting.yml [hunk 3]
```diff
@@ -33,24 +39,28 @@
...
-      setting_value_override: ['bar', 'foo']
+      setting_value: ['foo', 'bar']
       expected_change: false
 
   - include: tasks/setting.yml
     vars:
-      setting_name: content_action_accept_timeout
+      setting_name: token_duration
       setting_value: 99
+      setting_value_override: "99"
```
Comment (mdellweg) [inline tests/test_playbooks/setting.yml]: This is needed, because the 99 already arrives as a string in entity_dict. Maybe it is the jinja magic in tasks/setting.yaml, that converts it. ;(

File: plugins/modules/foreman_setting.py [hunk 4]
```diff
@@ -69,53 +67,45 @@
...
+    entity_dict['value'] = parameter_value_to_str(new_value, settings_type)
+    entity['value'] = parameter_value_to_str(entity['value'], settings_type)
 
-    entity_dict = sanitize_entity_dict(entity_dict, name_map)
+    changed, entity = module.ensure_entity('settings', entity_dict, entity, state='present', entity_spec=entity_spec)
 
-    changed, entity = naildown_entity(Setting, entity_dict, entity, 'present', module, check_type=True)
+    if entity:
+        # Fake the not serialized version as output
+        entity['value'] = new_value
```
Comment (evgeni) [inline plugins/modules/foreman_setting.py]: Why? When I `PUT` a new integer value (regardless whether the JSON I upload has it as a String or an Integer) the returned reply is an Integer, so it should compare correctly already?

```
# curl -X PUT -u admin:changeme -d '{"setting":{"value":361}}' -H 'Content-Type: application/json' -k https://foreman.example.com/api/v2/settings/132
{"value":361,"description":"Time in minutes installation tokens should be valid for, 0 to disable token generation","category":"Setting::Provisioning","settings_type":"integer","default":360,"created_at":"2019-07-16 06:55:35 UTC","updated_at":"2019-08-09 07:58:05 UTC","id":132,"name":"token_duration","full_name":"Token duration","category_name":"Provisioning"}

# curl -X PUT -u admin:changeme -d '{"setting":{"value":"362"}}' -H 'Content-Type: application/json' -k https://foreman.example.com/api/v2/settings/132
{"value":362,"description":"Time in minutes installation tokens should be valid for, 0 to disable token generation","category":"Setting::Provisioning","settings_type":"integer","default":360,"created_at":"2019-07-16 06:55:35 UTC","updated_at":"2019-08-09 07:59:13 UTC","id":132,"name":"token_duration","full_name":"Token duration","category_name":"Provisioning"}
```

**Comment by evgeni** [comment]
@sean797 do you plan/want to review this PR? Or is it good enough if I ACK it? :)

File: plugins/modules/foreman_setting.py [hunk 5]
```diff
@@ -69,53 +67,45 @@
...
+entity_spec = {
+    'name': {},
+    'value': {},
 }
 
 
 def main():
-    module = ForemanAnsibleModule(
+    module = ForemanApypieAnsibleModule(
         argument_spec=dict(
```
[the remark this answers is not in the document]
Comment (mdellweg) [reply plugins/modules/foreman_setting.py]: This one is special, as it has no `state`. Therefore it is not using the `EntityAnsibleModule`.
I don't know how it behaves with the `entity_spec`. Need to check that.

File: plugins/modules/foreman_setting.py (hunk 4 above)
Comment (mdellweg) [reply plugins/modules/foreman_setting.py]: The problem is not with integers:
`"msg": "Setting the setting failed! ('[\"foo\", \"bar\"]' != '[u'foo', u'bar']')"`

File: plugins/modules/foreman_setting.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mdellweg) [reply plugins/modules/foreman_setting.py]: At least in the current version, it is not supported to pass `entity_spec` to the non-entity-module. It sounds kind of reasonable, given the current naming.

File: plugins/modules/foreman_setting.py (hunk 4 above)
Comment (evgeni) [reply plugins/modules/foreman_setting.py]: Right. But it's the integers where it bites us in the tests. So my thought to update the `if` to:

```python
if entity and entity['settings_type'] in ['array', 'hash', 'yaml', 'json']:
  entity['value'] = new_value
```

File: plugins/modules/foreman_setting.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (evgeni) [reply plugins/modules/foreman_setting.py]: Right. (And maybe the name is wrong, but let's discuss that at another stage)

File: plugins/modules/foreman_setting.py [hunk 6]
```diff
@@ -69,53 +67,45 @@
...
-        entity_dict['value'] = entity.default or ""
+        entity_dict['value'] = entity['default'] or ''
+
+    settings_type = entity['settings_type']
+    new_value = entity_dict['value']
+    entity_dict['value'] = parameter_value_to_str(new_value, settings_type)
+    entity['value'] = parameter_value_to_str(entity['value'], settings_type)
 
-    entity_dict = sanitize_entity_dict(entity_dict, name_map)
+    changed, entity = module.ensure_entity('settings', entity_dict, entity, state='present', entity_spec=entity_spec)
```
[the remark this answers is not in the document]
Comment (evgeni) [reply plugins/modules/foreman_setting.py]: not here, not today.

File: plugins/modules/foreman_setting.py (hunk 4 above)
Comment (mdellweg) [reply plugins/modules/foreman_setting.py]: Hmm. It's more complicated than this, because the api wants `string` and gives us `integer`.
So in case, we did not update the setting, the only reasonable thing to return is the "unchanged" current entity. However we changed it already, to make it fit for the comparison in `_update_entity`. So the solution, i found, is to convert the input to integer, if the `settings_type` is `integer`.

File: tests/test_playbooks/tasks/setting.yml [hunk 7]
```diff
@@ -16,6 +16,8 @@
     that:
       - result.changed == expected_change
   when: expected_change is defined
+- debug:
```
Comment (evgeni) [inline tests/test_playbooks/tasks/setting.yml]: I think this can be removed again? :)

File: plugins/modules/foreman_setting.py (hunk 4 above)
Comment (evgeni) [reply plugins/modules/foreman_setting.py]: Works for me, thanks!

File: tests/test_playbooks/tasks/setting.yml (hunk 7 above)
Comment (mdellweg) [reply tests/test_playbooks/tasks/setting.yml]: Oh yes. :fallen_leaf:
